eclair-core/src/main/scala/fr/acinq/eclair/channel/InteractiveTxBuilder.scala

```diff
@@ -0,0 +1,722 @@
+    input1.previousTx.txid == input2.previousTx.txid && input1.previousTxOutput == input2.previousTxOutput
+  }
+
+  def toOutPoint(input: TxAddInput): OutPoint = OutPoint(input.previousTx, input.previousTxOutput.toInt)
+
+  def addRemoteSigs(fundingParams: InteractiveTxParams, partiallySignedTx: PartiallySignedSharedTransaction, remoteSigs: TxSignatures): Either[ChannelException, FullySignedSharedTransaction] = {
+    if (partiallySignedTx.tx.localInputs.length != partiallySignedTx.localSigs.witnesses.length) {
+      return Left(InvalidFundingSignature(fundingParams.channelId, Some(partiallySignedTx.tx.buildUnsignedTx())))
+    }
+    if (partiallySignedTx.tx.remoteInputs.length != remoteSigs.witnesses.length) {
+      return Left(InvalidFundingSignature(fundingParams.channelId, Some(partiallySignedTx.tx.buildUnsignedTx())))
+    }
+    val txWithSigs = FullySignedSharedTransaction(partiallySignedTx.tx, partiallySignedTx.localSigs, remoteSigs)
+    if (remoteSigs.txId != txWithSigs.signedTx.txid) {
+      return Left(InvalidFundingSignature(fundingParams.channelId, Some(partiallySignedTx.tx.buildUnsignedTx())))
+    }
+    // We allow a 5% error margin since witness size prediction could be inaccurate.
+    if (fundingParams.localAmount > 0.sat && txWithSigs.feerate < fundingParams.targetFeerate * 0.95) {
+      return Left(InvalidFundingFeerate(fundingParams.channelId, fundingParams.targetFeerate, txWithSigs.feerate))
+    }
+    val previousOutputs = {
+      val localOutputs = txWithSigs.tx.localInputs.map(i => toOutPoint(i) -> i.previousTx.txOut(i.previousTxOutput.toInt)).toMap
+      val remoteOutputs = txWithSigs.tx.remoteInputs.map(i => i.outPoint -> i.txOut).toMap
+      localOutputs ++ remoteOutputs
+    }
+    Try(Transaction.correctlySpends(txWithSigs.signedTx, previousOutputs, ScriptFlags.STANDARD_SCRIPT_VERIFY_FLAGS)) match {
+      case Failure(_) => Left(InvalidFundingSignature(fundingParams.channelId, Some(partiallySignedTx.tx.buildUnsignedTx()))) // NB: we don't send our signatures to our peer.
+      case Success(_) => Right(txWithSigs)
+    }
+  }
+
+}
+
+private class InteractiveTxBuilder(replyTo: ActorRef[InteractiveTxBuilder.Response],
+                                   fundingParams: InteractiveTxBuilder.InteractiveTxParams,
+                                   keyManager: ChannelKeyManager,
+                                   localParams: LocalParams,
+                                   remoteParams: RemoteParams,
+                                   commitTxFeerate: FeeratePerKw,
+                                   remoteFirstPerCommitmentPoint: PublicKey,
+                                   channelFlags: ChannelFlags,
+                                   channelConfig: ChannelConfig,
+                                   channelFeatures: ChannelFeatures,
+                                   wallet: OnChainChannelFunder,
+                                   previousAttempts: Seq[InteractiveTxBuilder.SignedSharedTransaction],
```

Comment:
Yes, that's exactly it. It can be inferred because it contains a `SignedSharedTransaction`, not a `PartiallySignedSharedTransaction`, which means both sides have signed (and thus broadcast), but this is a bit subtle.

Reply:
The naming is what confused me: I thought it was unsuccessful attempts. How about something like:
```suggestion
                                   previousTransactions: Seq[InteractiveTxBuilder.SignedSharedTransaction],
```

Reply:
Renamed in https://github.com/ACINQ/eclair/pull/2273/commits/6e0c15d430e3f4867e97196e7e350701753db688